internal/objectstore/remote/retriever.go

```diff
@@ -0,0 +1,392 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	"github.com/juju/clock"
+	jujuerrors "github.com/juju/errors"
+	"github.com/juju/worker/v4"
+	"github.com/juju/worker/v4/catacomb"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/internal/s3client"
+	internalworker "github.com/juju/juju/internal/worker"
+	"github.com/juju/juju/internal/worker/apiremotecaller"
+)
+
+const (
+	// NoRemoteConnections is returned when there are no remote connections
+	// available.
+	NoRemoteConnections = errors.ConstError("no remote connections available")
+
+	// BlobNotFound is returned when the requested blob is not found on any of
+	// the remote connections.
+	BlobNotFound = errors.ConstError("blob not found")
+)
+
+// BlobsClient is an interface for retrieving objects from an object store.
+type BlobsClient interface {
+	// GetObject returns a reader for the object with the given key in the
+	// given bucket.
+	GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, error)
+}
+
+// NewObjectClientFunc is a function that creates a new BlobsClient.
+type NewObjectClientFunc func(url string, client s3client.HTTPClient, logger logger.Logger) (BlobsClient, error)
+
+// BlobRetriever is responsible for retrieving blobs from remote API servers.
+type BlobRetriever struct {
+	catacomb catacomb.Catacomb
+
+	namespace string
+
+	apiRemoteCallers apiremotecaller.APIRemoteCallers
+	newObjectClient  NewObjectClientFunc
```

Comment:
Why do we need a factory? wouldn't a client injected be enough?
```suggestion
	blobsClient  BlobsClient
```

Reply:
You can't use 1 client to speak to multiple backends. It's a limitation of the clients.